# Importing Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# importing my customer datasets.csv as df_customer
df_customer = pd.read_csv(r'C:\Users\Babafemi Olayinka\Jupyter Notebook\FEB-2023 INSTACART BASKET ANALYSIS\Data\Original Data\customers.csv' , index_col = False)

In [3]:
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# Data Wrangling for my customer df
 I renamed some columns, dropped some columns and check data types for the columns

In [4]:
# checking the dimension of my customer df
df_customer.shape

(206209, 10)

In [5]:
# importing my df_new_products as a pickle file as my latest entire df before visuals.
df_new_products = pd.read_pickle(r'C:\Users\Babafemi Olayinka\Jupyter Notebook\FEB-2023 INSTACART BASKET ANALYSIS\Data\Prepared Data\df_orders_merged_products_grouped.pkl')

In [7]:
df_new_products.()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  ...  \
0   3139998      138            28          6                 11  ...   
1   1977647      138            30          6                 17  ...   
2    389851      709             2          0                 21  ...   
3    652770      764             1          3                 13  ...   
4   1813452      764             3          4                 17  ...   

      Busiest day  Busiest days  busiest_period_of_day max_order  \
0  Regularly busy  Regular days            Most orders        32   
1  Regularly busy  Regular days         Average orders        32   
2     Busiest day  Busiest days         Average orders         5   
3  Regularly busy  Slowest days            Most orders         3   
4      Least busy  Slowest days         Average orders         3   

       loyalty_flag max_price average_price spending_flag  order_freq  \
0  Regular customer      20.0          20.0  High spender         8.0   
1  Regular customer      20.0          20.0  High spender         8.0   
2      New customer      18.5          18.5  High spender         7.0   
3      New customer      14.0          14.0  High spender         9.0   
4      New customer      14.0          14.0  High spender         9.0   

     order_freq_flag  
0  frequent customer  
1  frequent customer  
2  frequent customer  
3  frequent customer  
4  frequent customer  

[5 rows x 25 columns]

In [8]:
# dropping my 'First Name and Surnam' columns in my customer df
df_customer = df_customer.drop(columns = ['First Name', 'Surnam'])

In [9]:
# Finding missing values in my customer df. There are no missing values
df_customer.isnull().sum()

user_id         0
Gender          0
STATE           0
Age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

In [17]:
# Renaming some columns within the customer df to a more distinct name
df_customer.rename(columns = {'STATE' : 'State', 'n_dependants' : 'Number_Of_Dependents', 'fam_status' : 'Family_Status'}, inplace = True)

In [19]:
df_customer.head()

user_id  Gender       State  Age date_joined  No_Of_Dependents  \
0   26711  Female    Missouri   48    1/1/2017                 3   
1   33890  Female  New Mexico   36    1/1/2017                 0   
2   65803    Male       Idaho   35    1/1/2017                 2   
3  125935  Female        Iowa   40    1/1/2017                 0   
4  130797  Female    Maryland   26    1/1/2017                 1   

  Family_Status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [14]:
# changing the data type fron int64 to str for my user_id column only.
df_customer['user_id'] = df_customer['user_id'].astype('str')

In [15]:
# Telling python to return the data type of user_i incustomer df, and it return as 'o' stands for object
df_customer['user_id'].dtype

dtype('O')

# Data consistency check for my customer df
I checked for any discrepancies in the descriptive statistics of my df, checked if any columns also have duplicated values  for all the rows within my dataframe, i also checked whether my dataframw contains any mixed-type columns using the FOR-LOOP fx.

In [20]:
# using the describe() to check for descriptive statistics for any irregularies 
df_customer.describe()

Age  No_Of_Dependents         income
count  206209.000000     206209.000000  206209.000000
mean       49.501646          1.499823   94632.852548
std        18.480962          1.118433   42473.786988
min        18.000000          0.000000   25903.000000
25%        33.000000          0.000000   59874.000000
50%        49.000000          1.000000   93547.000000
75%        66.000000          3.000000  124244.000000
max        81.000000          3.000000  593901.000000

## Everything looks good for the Age and no_of_dependants. 
The mean age is 49, the max age is 81 and min age is 18. The mean and median age is the same showing the a symmetric distribution of the data. The max dependants for a family is 3 and min is 0, median is 1. The mean of the income is 94632, the median is 93547, meaning the excessive values which may stand as the outliers as dragged the mean value toward themselves.


In [21]:
df_new_products['user_id'] = df_new_products['user_id'].astype('str')

In [23]:
# changing my datatype of the latest merged df in previuos exercise
df_new_products['user_id'].dtype

dtype('O')

In [24]:
# checking for duplicated values within the columns in my enrite df. There are no duplicated values within my df
df_customer[df_customer.duplicated()]

Empty DataFrame
Columns: [user_id, Gender, State, Age, date_joined, No_Of_Dependents, Family_Status, income]
Index: []

In [34]:
# checking with the fx FOR-LOOP to see if any of my columns contain mixed datatype within my df.
for col in df_customer.columns.tolist():
  weird = (df_customer[[col]].applymap(type) != df_customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customer[weird]) > 0:
    print (col)

# Merging my 2 different df (df_customer and df_new_products) 
I will be using fx merge because the 2 dataframes do not have same shape
You may noticed i did some coding twice, this is because my computer shut down due to shaortage of space, i have to clean some stuff out for more space. 
I have also created a new subset which contain only 2 million rows of my original df. And this will be used to merge the customer dataframe below. Also that is what i will keep to for the rest of my task and visuals.

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [3]:
df_new_products = pd.read_pickle(r'C:\Users\Babafemi Olayinka\Jupyter Notebook\FEB-2023 INSTACART BASKET ANALYSIS\Data\Prepared Data\df_orders_merged_products_grouped.pkl')

In [6]:
df_new_products_sub = df_new_products[:2000000]

In [9]:
# Due to space, have created a subset of my my original data df_new_products. my new df will have only 2 million rows.
df_new_products_sub.shape

(2000000, 25)

In [13]:
# dropping my 'First Name and Surnam' columns in my customer df
df_customer = df_customer.drop(columns = ['First Name', 'Surnam'])

In [14]:
df_customer.head()

user_id  Gender       STATE  Age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

In [15]:
# Renaming some columns within the customer df to a more distinct name
df_customer.rename(columns = {'STATE' : 'State', 'n_dependants' : 'Number_Of_Dependents', 'fam_status' : 'Family_Status'}, inplace = True)

In [16]:
df_customer.head()

user_id  Gender       State  Age date_joined  Number_Of_Dependents  \
0    26711  Female    Missouri   48    1/1/2017                     3   
1    33890  Female  New Mexico   36    1/1/2017                     0   
2    65803    Male       Idaho   35    1/1/2017                     2   
3   125935  Female        Iowa   40    1/1/2017                     0   
4   130797  Female    Maryland   26    1/1/2017                     1   

  Family_Status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [10]:
df_customer = pd.read_csv(r'C:\Users\Babafemi Olayinka\Jupyter Notebook\FEB-2023 INSTACART BASKET ANALYSIS\Data\Original Data\customers.csv' , index_col = False)

In [17]:
df_prods_ords_customer = pd.merge(df_new_products_sub, df_customer, on = 'user_id', how = 'outer')

In [18]:
df_prods_ords_customer.head()

product_id                product_name  aisle_id  department_id  prices  \
0         1.0  Chocolate Sandwich Cookies      61.0           19.0     5.8   
1         1.0  Chocolate Sandwich Cookies      61.0           19.0     5.8   
2       907.0        Premium Sliced Bacon     106.0           12.0    20.0   
3       907.0        Premium Sliced Bacon     106.0           12.0    20.0   
4      1000.0                    Apricots      18.0           10.0    12.9   

    order_id  user_id  order_number  order_dow  order_hour_of_day  ...  \
0  3139998.0      138          28.0        6.0               11.0  ...   
1  1977647.0      138          30.0        6.0               17.0  ...   
2  3160996.0      138           1.0        5.0               13.0  ...   
3  2254091.0      138          10.0        5.0               14.0  ...   
4   505689.0      138           9.0        6.0               12.0  ...   

   spending_flag  order_freq    order_freq_flag Gender      State Age  \
0   High spender         8.0  frequent customer   Male  Minnesota  81   
1   High spender         8.0  frequent customer   Male  Minnesota  81   
2   High spender         8.0  frequent customer   Male  Minnesota  81   
3   High spender         8.0  frequent customer   Male  Minnesota  81   
4   High spender         8.0  frequent customer   Male  Minnesota  81   

  date_joined Number_Of_Dependents  Family_Status income  
0    8/1/2019                    1        married  49620  
1    8/1/2019                    1        married  49620  
2    8/1/2019                    1        married  49620  
3    8/1/2019                    1        married  49620  
4    8/1/2019                    1        married  49620  

[5 rows x 32 columns]

In [19]:
# checking with the frequency to see if it full join the 2 df. It full join.
df_prods_ords_customer['_merge'].value_counts()

both          2000000
left_only           0
right_only          0
Name: _merge, dtype: int64

In [20]:
# verifying the dimension of my new merged df. 
df_prods_ords_customer.shape

(2027234, 32)

# Exporting my newlymerged customer , products and orders dataset as 
df_prods_ords_customer with just first 2 million rows 

In [21]:
path = r'C:\Users\Babafemi Olayinka\Jupyter Notebook\FEB-2023 INSTACART BASKET ANALYSIS'

In [22]:
path

'C:\\Users\\Babafemi Olayinka\\Jupyter Notebook\\FEB-2023 INSTACART BASKET ANALYSIS'

In [23]:
df_prods_ords_customer.to_pickle(os.path.join(path, 'Data','Prepared Data', 'df_prods_ords_customer.pkl'))